Zend/zend_weakrefs.c

```diff
@@ -144,6 +143,22 @@ static void zend_weakref_unregister(zend_object *object, void *payload) {
 		ZEND_WEAKREF_GET_PTR(payload), ZEND_WEAKREF_GET_TAG(payload), obj_addr);
 }
 
+ZEND_API zval *zend_weakrefs_hash_add(HashTable *ht, zend_object *key, zval *pData) {
+	zval *zv = zend_hash_index_add(ht, (zend_ulong) key, pData);
+	if (zv) {
+		zend_weakref_register(key, ZEND_WEAKREF_ENCODE(ht, ZEND_WEAKREF_TAG_MAP));
+	}
+	return zv;
+}
+
+ZEND_API zend_result zend_weakrefs_hash_del(HashTable *ht, zend_object *key) {
+	zend_result result = zend_hash_index_del(ht, (zend_ulong) key);
+	if (result == SUCCESS) {
+		zend_weakref_unregister(key, ZEND_WEAKREF_ENCODE(ht, ZEND_WEAKREF_TAG_MAP));
```

Comment:
zend_weakref_unregister already does the hash del, so it might make sense to replace your hash_index_del here with hash_index_exists.